aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -42,383 +47,506 @@ module aptos_framework::fungible_asset {
+    public fun wallet_exists_at(wallet: address): bool {
+        exists<FungibleAssetWallet>(wallet)
     }
 
     /// Return the underlying metadata object
     public fun metadata_from_asset(fa: &FungibleAsset): Object<FungibleAssetMetadata> {
         fa.metadata
     }
 
+    #[view]
     /// Return the underlying metadata object.
-    public fun metadata_from_wallet(
-        wallet: &Object<FungibleAssetWallet>
-    ): Object<FungibleAssetMetadata> acquires FungibleAssetWallet {
-        borrow_fungible_asset(wallet).metadata
+    public fun wallet_metadata<T: key>(wallet: Object<T>): Object<FungibleAssetMetadata> acquires FungibleAssetWallet {
+        borrow_wallet_resource(&wallet).metadata
     }
 
-    /// Return `amount` inside.
+    /// Return `amount` of a given fungible asset.
     public fun amount(fa: &FungibleAsset): u64 {
         fa.amount
     }
 
-    /// Destroy `wallet` object.
-    public(friend) fun destory_fungible_asset_wallet(
-        wallet: Object<FungibleAssetWallet>
-    ) acquires FungibleAssetWallet {
-        let FungibleAssetWallet {
-            metadata: _,
-            balance: _,
-            allow_ungated_transfer: _,
-            delete_ref
-        } = move_from<FungibleAssetWallet>(object::object_address(&wallet));
-        object::delete(delete_ref);
+    #[view]
+    /// Get the balance of a given wallet.
+    public fun balance<T: key>(wallet: Object<T>): u64 acquires FungibleAssetWallet {
+        if (wallet_exists_at(object::object_address(&wallet))) {
+            borrow_wallet_resource(&wallet).balance
+        } else {
+            0
+        }
+    }
+
+    #[view]
+    /// Return whether a wallet can freely send or receive fungible assets.
+    /// If the wallet has not been created, we default to returning true as deposits can be sent to it.
+    public fun ungated_transfer_allowed<T: key>(wallet: Object<T>): bool acquires FungibleAssetWallet {
+        !wallet_exists_at(object::object_address(&wallet)) ||
+            borrow_wallet_resource(&wallet).allow_ungated_transfer
+    }
+
+    public fun asset_metadata(fa: &FungibleAsset): Object<FungibleAssetMetadata> {
+        fa.metadata
+    }
+
+    /// Get the underlying metadata object from `MintRef`.
+    public fun mint_ref_metadata(ref: &MintRef): Object<FungibleAssetMetadata> {
+        ref.metadata
+    }
+
+    /// Get the underlying metadata object from `TransferRef`.
+    public fun transfer_ref_metadata(ref: &TransferRef): Object<FungibleAssetMetadata> {
+        ref.metadata
+    }
+
+    /// Get the underlying metadata object from `BurnRef`.
+    public fun burn_ref_metadata(ref: &BurnRef): Object<FungibleAssetMetadata> {
+        ref.metadata
+    }
+
+    /// Transfer `amount` of fungible metadata of `metadata` to `receiver`.
+    /// Note: it does not move the underlying object.
+    public entry fun transfer<T: key>(
+        sender: &signer,
+        from_wallet: Object<T>,
```

Comment:
Why isn't it `Object<FungibleAssetWallet`? What else can it be?

Reply:
Updated